Nested groups with `field-list` appearance

grzesiek2010 opened the pull request:
Closes #1255 

#### What has been done to verify that this works as intended?
I tested lots of cases and I prepared an automated test.

#### Why is this the best possible solution? Were any other approaches considered?
We should allow adding groups inside a field-list group. I mean regular groups not repeat sections (it doesn't make sense), that's why I used a code for recursively grabbing all indices in a field-list group. This is the only solution I can imagine here.

#### Are there any risks to merging this code? If so, what are they?
In my opinion, it's very risky. I tried to be careful and I tested lot's of cases, I prepared an automated test too but you can never be sure everything is ok. It's possible I missed something. That's why I think that navigation in forms should be really really well tested by @mmarciniak90  

#### Do we need any specific form for testing your changes? If so, please attach one.
[form1.xml.txt](https://github.com/opendatakit/collect/files/1681571/form1.xml.txt)
[form2.xml.txt](https://github.com/opendatakit/collect/files/1681572/form2.xml.txt)
[form3.xml.txt](https://github.com/opendatakit/collect/files/1681573/form3.xml.txt)
[form4.xml.txt](https://github.com/opendatakit/collect/files/1681574/form4.xml.txt)
[form5.xml.txt](https://github.com/opendatakit/collect/files/1681575/form5.xml.txt)
[form6.xml.txt](https://github.com/opendatakit/collect/files/1681576/form6.xml.txt)
[form7.xml.txt](https://github.com/opendatakit/collect/files/1681577/form7.xml.txt)
[form8.xml.txt](https://github.com/opendatakit/collect/files/1681578/form8.xml.txt)
[form9.xml.txt](https://github.com/opendatakit/collect/files/1681579/form9.xml.txt)

grzesiek2010 committed:
Implemented required changes

grzesiek2010 committed:
Improved the comment

grzesiek2010 committed:
Added tests

grzesiek2010 committed:
Added comments

grzesiek2010 committed:
Fixed 'unchecked assignment java util array' warning

grzesiek2010 commented:
If it's ok we should create a separate issue for visual improvements because we need to decide how to highlight nested group names.

mmarciniak90 commented:
I noticed one case:
* User starts form: [metadata2.xml.txt](https://github.com/opendatakit/collect/files/1689484/metadata2.xml.txt)

Results:
* Error message is visible on Collect v1.12.2
<img height="40%" width="40%" src="https://user-images.githubusercontent.com/3907554/35735525-12da6c9c-081d-11e8-9073-3a2dc0b86891.png">
* Loading form does not end on Collect with this changes

grzesiek2010 commented:
@mmarciniak90 
I tried v1.12.2 and current master branch and I can't see such a dialog. Please check other devices. If it's not a problem caused by my changes we should create a separate issue.

mmarciniak90 commented:
Yes, @grzesiek2010  you have right. This problem depends on Android version. 
Error dialog is visible on Android 4.4 and below.
Loading form does not end on Android 5.1 and above.
I created a separate issue #1833

mmarciniak90 commented:
Tested with success!
Verified on Android: 4.1, 4.4, 5.1, 6.0, 7.0

@opendatakit-bot label "behavior verified"

dcbriccetti commented on collect_app/src/main/java/org/odk/collect/android/logic/FormController.java:
```diff
@@ -399,8 +399,15 @@ private boolean indexIsInFieldList(FormIndex index) {
+            // If at least one of groups you are inside is field list your index is in field list
+            boolean groupIsFieldList;
+            for (FormEntryCaption caption : captions) {
+                groupIsFieldList = groupIsFieldList(caption.getIndex());
```
I suggest inlining `groupIsFieldList`. You’ll remove two lines of code and lose no clarity.

dcbriccetti commented on collect_app/src/main/java/org/odk/collect/android/logic/FormController.java:
```diff
@@ -570,13 +569,12 @@ public int stepToPreviousScreenEvent() throws JavaRosaException {
+                            boolean groupIsFieldList;
+                            for (FormEntryCaption caption : fclist) {
+                                groupIsFieldList = groupIsFieldList(caption.getIndex());
+                                if (groupIsFieldList) {
```
Inline

dcbriccetti commented on collect_app/src/main/java/org/odk/collect/android/logic/FormController.java:
```diff
@@ -812,7 +810,7 @@ public void setLanguage(String language) {
     public FormEntryPrompt[] getQuestionPrompts() throws RuntimeException {
 
-        ArrayList<FormIndex> indicies = new ArrayList<FormIndex>();
+        ArrayList<FormIndex> indicies = new ArrayList<>();
```
Spelling, and it’s a good practice to make the type general: `List` in this case.

dcbriccetti commented on collect_app/src/main/java/org/odk/collect/android/logic/FormController.java:
```diff
@@ -880,6 +853,23 @@ public void setLanguage(String language) {
         return questions;
     }
 
+    private List<FormIndex> getIndices(GroupDef gd, FormIndex idxChild) {
```
I wonder if this method could be made easier to read and understand. I offer some suggestions. 

![screen shot 2018-02-07 at 4 01 20 pm](https://user-images.githubusercontent.com/55099/35948236-804cffc2-0c20-11e8-9c23-cebb89f7dfeb.png)

It might be clearer still if you didn’t modify the formal parameter idxChild. And even though this is a private method, a doc comment might be warranted.

dcbriccetti commented on collect_app/src/main/java/org/odk/collect/android/logic/FormController.java:
```diff
@@ -503,18 +510,14 @@ public int stepToNextEvent(boolean stepIntoGroup) {
-            // don't descend
-            idxChild = formEntryController.getModel().incrementIndex(idxChild, false);
-        }
+        indicies.addAll(getIndices(gd, idxChild));
```
Why add the result of getIndices to a second List? I suggest a simplification. Note the correct spelling of *indices*, and use of the more general `List` instead of `ArrayList`.

![screen shot 2018-02-07 at 3 19 08 pm](https://user-images.githubusercontent.com/55099/35946921-7d70343c-0c1a-11e8-9711-a11f2a1435ef.png)

dcbriccetti commented:
Now I’ll get in and really understand the change, and may have more comments.

dcbriccetti commented on collect_app/src/androidTest/java/org/odk/collect/android/FormNavigationTestCase.java:
```diff
@@ -0,0 +1,390 @@
+        formLoaderTask.execute(formPath(formName));
+    }
+
+    @Test
```
There are 390 lines of code spent in this file. How about factoring out the identical parts of these tests, or combining them in some other way.

“My point today is that, if we wish to count lines of code, we should not regard them as "lines produced" but as "lines spent": the current conventional wisdom is so foolish as to book that count on the wrong side of the ledger.”
Dijkstra (1988) "On the cruelty of really teaching computing science (EWD1036).

dcbriccetti reviewed (changes requested):
This concludes phase one of my review. I’ll return after FormController.getIndices is easier to understand.

grzesiek2010 committed:
Code improvements

grzesiek2010 committed:
Refactored the test

grzesiek2010 commented:
@dcbriccetti Thanks for your review, I implemented all code improvements you suggested and refactored the test too.

grzesiek2010 committed:
Code improvement

grzesiek2010 commented:
I'm adding @opendatakit-bot label "behavior verified" again since removing it was a mistake.

lognaturel commented on collect_app/src/androidTest/assets/forms/formNavigationTestForms/form1.xml:
```diff
@@ -0,0 +1,61 @@
+<h:html xmlns="http://www.w3.org/2002/xforms" xmlns:h="http://www.w3.org/1999/xhtml" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:jr="http://openrosa.org/javarosa">
+    <h:head>
+        <h:title>form1</h:title>
```
Name something like group-bind-no-ref-in-fieldlist

lognaturel commented on collect_app/src/androidTest/assets/forms/formNavigationTestForms/form2.xml:
```diff
@@ -0,0 +1,61 @@
+        <h:title>form2</h:title>
+        <model>
+            <instance>
+                <data id="build_form2_1517303824">
```
Name something like field-list-in-field-list

lognaturel commented on collect_app/src/androidTest/assets/forms/formNavigationTestForms/form3.xml:
```diff
@@ -0,0 +1,61 @@
+<h:html xmlns="http://www.w3.org/2002/xforms" xmlns:h="http://www.w3.org/1999/xhtml" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:jr="http://openrosa.org/javarosa">
+    <h:head>
+        <h:title>form3</h:title>
```
Name something like group-no-bind-no-ref-in-field-list

lognaturel commented on collect_app/src/androidTest/assets/forms/formNavigationTestForms/form4.xml:
```diff
@@ -0,0 +1,61 @@
+        <h:title>form4</h:title>
+        <model>
+            <instance>
+                <data id="build_form4_1517307219">
```
group-no-bind-no-ref-in-group

lognaturel commented on collect_app/src/androidTest/assets/forms/formNavigationTestForms/form5.xml:
```diff
@@ -0,0 +1,61 @@
+        <h:title>form5</h:title>
+        <model>
+            <instance>
+                <data id="build_form4_1517307426">
```
group-no-bind-no-ref-after-text-in-group

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/logic/FormController.java:
```diff
@@ -570,13 +567,10 @@ public int stepToPreviousScreenEvent() throws JavaRosaException {
                         if (ODKView.FIELD_LIST.equalsIgnoreCase(gd.getAppearanceAttr())) {
                             // OK this group is a field-list... see what the parent is...
```
Replace with `// jump to outermost containing field-list`

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/logic/FormController.java:
```diff
@@ -812,7 +806,7 @@ public void setLanguage(String language) {
     public FormEntryPrompt[] getQuestionPrompts() throws RuntimeException {
 
-        ArrayList<FormIndex> indicies = new ArrayList<FormIndex>();
+        List<FormIndex> indices = new ArrayList<>();
```
Should be declared when `getIndicesForForGroup` is first called so you don't need the `addAll`.

dcbriccetti replied:
Yes, indeed.
![screen shot 2018-02-18 at 6 47 10 pm](https://user-images.githubusercontent.com/55099/36360878-7fae6834-14dc-11e8-99b2-b1aa6653c5e5.png)

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/logic/FormController.java:
```diff
@@ -880,6 +849,24 @@ public void setLanguage(String language) {
         return questions;
     }
 
+    private List<FormIndex> getIndicesForGroup(GroupDef gd, FormIndex idxChild) {
```
A comment here would be helpful. Something like `Recursively get all indices contained in this group and its children.`

lognaturel replied:
I would tend to split this into a method that only takes a `GroupDef` as a parameter and a recursive helper that takes both the group and the index (`currentChild` would be a more descriptive name).

dcbriccetti replied:
I wanted to work out what you meant here, @lognaturel. Like this?

![screen shot 2018-02-19 at 1 04 49 pm](https://user-images.githubusercontent.com/55099/36396983-83633aba-1575-11e8-8be9-def5843cbe37.png)

lognaturel replied:
That’s what I had in mind. Worth it or just personal preference? I also don’t think idxChild is a very helpful name and would tend to use currentChild or something that doesn’t require thinking.

dcbriccetti replied:
I think it’s worth it. I also like currentChild (or idxCurrentChild or currentChildIndex), but forgot to make that change in the code shown above.

dcbriccetti replied:
It removes code duplication, and the concern of finding the starting index, from the callers.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/logic/FormController.java:
```diff
@@ -824,41 +818,16 @@ public void setLanguage(String language) {
-                // If it is, then silently recurse into it to pick up its elements.
-                // Work-around for the inconsistent treatment of field-list groups and repeats in
-                // 1.1.7 that
-                // either breaks forms generated by build or breaks forms generated by XLSForm.
```
@grzesiek2010 Do you understand what broken state this was referring to? The commit that removes this comment should have a description of why it is no longer relevant. It might also be helpful to include some tests that use forms from XLSForm. The only difference I know of is that XLSForm includes `ref` attributes on groups whereas Build doesn't but there may be something else.

grzesiek2010 replied:
It was possible to have a field-list group inside another field-list but only if it was the only child. I tested such a form as I remember. It was just one case when you were able to have another group inside a field-list before my changes. So we just handled the simplest case as I understand.

grzesiek2010 replied:
Since I implemented changes for handling everything we don't need that.

grzesiek2010 replied:
It was this case https://github.com/grzesiek2010/collect/blob/559b87a4d1ac23550a7d214ca6b36f561b97bf10/collect_app/src/androidTest/assets/forms/formNavigationTestForms/fieldListInFieldList.xml

lognaturel replied:
This does not answer the question of WHY the field-list in field-list case was explicitly handled. The comment makes it seem like that was a deliberate choice and that has me nervous.

grzesiek2010 replied:
So I don't know why. I guess it was the easiest case and someone wanted to handle that (at least that). Now we want to handle all possible cases. But I don't really understand the last line.

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/logic/FormController.java:
```diff
@@ -824,41 +818,16 @@ public void setLanguage(String language) {
-            for (int i = 0; i < indicies.size(); i++) {
-                FormIndex index = indicies.get(i);
+            List<FormEntryPrompt> questionList = new ArrayList<>();
+            for (int i = 0; i < indices.size(); i++) {
```
While you're here, change to `for (FormIndex index : indices)`?

lognaturel reviewed (dismissed):
Nice, @grzesiek2010! This is complex code. The PR template question about the best possible solution is a good opportunity to consider the way the code was written. For example, in this case, is recursively grabbing all indices in a group the only approach? What else might you have considered? Why is it a necessary change?

It's great to see some tests as well. Could you please name the XML files according to the cases that they test? I made some suggestions. It would also be helpful to have a comment in the test file describing why these tests are valuable and what issue they relate to. Consider also making clear which cases failed before your implemented changes and why. Finally, when there is a failure it should be clear which of the forms has failed. Currently it's not clear because we always get a comparison error in the same `testIndices` method.

@dcbriccetti Anything else from your side?

dcbriccetti commented on collect_app/src/androidTest/java/org/odk/collect/android/FormNavigationTestCase.java:
```diff
@@ -44,315 +44,60 @@
 
     @Test
-    public void test1() throws IOException {
+    public void formNavigationTestCase() throws IOException {
```
It’s great to see this so much shorter. Thanks.

lognaturel replied:
Yes but now it's not possible to tell what case failed! I think this is a good opportunity for parameterized tests.

lognaturel replied:
@grzesiek2010 do you see what I mean by this? For example, change any of the `"-1, "` entries to `"-10"`. How will you know which case failed? 

You can read about parameterized tests at https://github.com/junit-team/junit4/wiki/parameterized-tests and @ggalmazor will surely have some good advice if you need.

Good to see a comment at the top! "it might be tricky" should be elaborated on -- what are the kinds of errors or regressions this verifies? Tests can be a big help when getting to know new code so it's a good opportunity to provide a little context. What do the lists of numbers represent? Why do you go backwards once at the end while navigating the form? Etc.

grzesiek2010 replied:
I implemented changes using parameterized tests let me know if it is satisfying for you.

grzesiek2010 replied:
But there is one problem when any test failed it displays a name of the last one...I'm not sure why it was ok when I tested an example from your link. I guess it's because we load a form and it's another task.

grzesiek2010 commented:
@lognaturel I think it's ready.

dcbriccetti commented on collect_app/src/main/java/org/odk/collect/android/logic/FormController.java:
```diff
@@ -842,21 +837,22 @@ public void setLanguage(String language) {
     }
 
-    private List<FormIndex> getIndices(GroupDef gd, FormIndex idxChild) {
+    // Recursively get all indices contained in this group and its children
```
Might as well make this a doc comment and change the verb mood to declarative: “gets”.
/** Recursively gets all indices contained in this group and its children */

grzesiek2010 committed:
Improved the FormNavigationTestCase

dcbriccetti commented on collect_app/src/androidTest/java/org/odk/collect/android/FormNavigationTestCase.java:
```diff
@@ -0,0 +1,139 @@
+
+/**
+ * This test has been created in order to check indices while navigating through a form.
+ * It's especially important while navigate through a form which contains nested groups and if we
```
while navigating through a form that contains

dcbriccetti commented on collect_app/src/androidTest/java/org/odk/collect/android/FormNavigationTestCase.java:
```diff
@@ -0,0 +1,139 @@
+ * It's especially important while navigate through a form which contains nested groups and if we
+ * use groups with field-list appearance because in that case we need to collect all indices of
+ * questions we want to display on one page (we need to recursively get all indices contained in
+ * such a group and its children). It might be also tricky when navigate backwards because then we
```
navigating

dcbriccetti commented on collect_app/src/androidTest/java/org/odk/collect/android/FormNavigationTestCase.java:
```diff
@@ -0,0 +1,139 @@
+            @Override
+            public void loadingComplete(FormLoaderTask task) {
+                try {
+                    for (int i = 0; i < expectedIndices.length - 1; i++) {
```
I’m finding it challenging to understand the “- 1”, “- 2”, “stepToNext...”, “stepToPrevious...”, “+ 1”, and why there are two assertEquals(expectedIndices...) each time through the loop. I wonder if a few comments in here would help me and others.

grzesiek2010 replied:
We have two asserts because I check the current index and the index after navigating (back or forward). I added a comment.

lognaturel replied:
I agree this is still confusing. How about a comment on `testIndices` that says something like "For each form, simulate swiping forward through screens until the end of the form and then swiping back once. Verify the expected indices before and after each swipe."

And for the test patterns, something above the array like "Expected indices when swiping forward until the end of the form and back once. An index of `-1` indicates the start or end of a form."

I don't really know how/if to explain what the `FormIndex` toString represents (comma-separated lists). A `FormIndex` represents where in the form we are. The `toString` prints out the path through nested structures that gets us to the current screen. The idea is that a field-list should not have any nesting. That's why they show up as "0". The tests demonstrate that this is true even when other groups are nested in the field list. Something like "0, 0, 2" means it's the 3rd question within two levels of nesting, both groups being the first at their level.

I think it's ok to just add the two comments I mentioned above. @dcbriccetti?

grzesiek2010 replied:
I've just added comments.

dcbriccetti commented on collect_app/src/androidTest/java/org/odk/collect/android/FormNavigationTestCase.java:
```diff
@@ -0,0 +1,139 @@
+        formLoaderTask.execute(formPath(formName));
+    }
+
+    private void prepareFile(String formName) throws IOException {
```
Pardon my Android ignorance, but why must we copy these files? Perhaps the method name, or a comment, could say why.

grzesiek2010 replied:
It's because FormLoaderTask loads forms from SD card so we need to put all forms there. I added a comment.

grzesiek2010 committed:
Fixed problem with AsyncTask

grzesiek2010 committed:
Improvements

grzesiek2010 commented:
@lognaturel I fixed the problem with my test: https://github.com/opendatakit/collect/pull/1821/commits/fa52abe20f94f4a29f972a1f1089919d1d01a98a 
As I thought it was because of used AsyncTask.
I added [AsynctTask().get()](https://developer.android.com/reference/android/os/AsyncTask.html#get%28%29) and now I know which set of parameters failed.

lognaturel reviewed (commented):
It's still a bit hard to tell where a failure has occurred because all forms have multiple -1 entries and some have other duplicates but at least it's clear which form is failing so I'm satisfied. I've suggested a couple more comments to add. @dcbriccetti, please mark as `needs testing` when you're ok with this.

grzesiek2010 committed:
Added comments

lognaturel commented:
@dcbriccetti Have the changes you requested been satisfyingly addressed?

dcbriccetti commented on collect_app/src/androidTest/java/org/odk/collect/android/FormNavigationTestCase.java:
```diff
@@ -0,0 +1,149 @@
+    @Parameters(name = "{0}")
+    public static Iterable<Object[]> data() {
+        // Expected indices when swiping forward until the end of the form and back once.
+        // An index of -1 indicates the start or end of a form
```
A single comment with no period is fine. But when there are two sentences together they should both end with periods.

dcbriccetti commented on collect_app/src/androidTest/java/org/odk/collect/android/FormNavigationTestCase.java:
```diff
@@ -0,0 +1,149 @@
+            public void loadingComplete(FormLoaderTask task) {
+                try {
+                    // For each form, simulate swiping forward through screens until the end of the
+                    // form and then swiping back once. Verify the expected indices before and after each swipe
```
The second sentence needs a period.

dcbriccetti commented on collect_app/src/androidTest/java/org/odk/collect/android/FormNavigationTestCase.java:
```diff
@@ -0,0 +1,149 @@
+    /*
+    FormLoaderTask loads forms from SD card so we need to put each form there
+     */
+    private void prepareFile(String formName) throws IOException {
```
This could be a doc comment ( /** ... */ ). Might the method have a more descriptive, specific name? `copyToSdCard`, perhaps?

dcbriccetti commented on collect_app/src/main/java/org/odk/collect/android/logic/FormController.java:
```diff
@@ -399,8 +400,13 @@ private boolean indexIsInFieldList(FormIndex index) {
             }
-            FormEntryCaption grp = captions[captions.length - 2];
-            return groupIsFieldList(grp.getIndex());
+            // If at least one of groups you are inside is field list your index is in field list
```
```
If at least one of     groups you are inside is   field list  your index is in   field list
If at least one of the groups you are inside is a field list, your index is in a field list
```

dcbriccetti commented on collect_app/src/androidTest/java/org/odk/collect/android/FormNavigationTestCase.java:
```diff
@@ -0,0 +1,149 @@
+                {"fieldListWithQuestionsAndRegularGroupsInside.xml", new String[] {"-1, ", "0, ", "-1, ", "0, "}},
+                {"threeNestedFieldListGroups.xml", new String[] {"-1, ", "0, ", "-1, ", "0, "}}
+        });
+    }
```
You might consider a further beautification like this, to remove some boilerplate and line things up better. If you prefer it how you have it now, that’s fine with me. This is inspired by domain-specific languages.
```
@Parameters(name = "{0}")
public static Iterable<Object[]> data() {
    // Expected indices when swiping forward until the end of the form and back once.
    // An index of -1 indicates the start or end of a form
    return Arrays.asList(
            ei("simpleFieldList.xml",
                    "-1, ", "0, ", "-1, ", "0, "),
            ei("fieldListInFieldList.xml",
                    "-1, ", "0, ", "-1, ", "0, "),
            ei("regularGroupWithFieldListGroupInside.xml",
                    "-1, ", "0, 0, ", "-1, ", "0, 0, "),
            ei("twoNestedRegularGroups.xml",
                    "-1, ", "0, 0, 0, ", "0, 0, 1, ", "0, 0, 2, ", "-1, ", "0, 0, 2, "),
            ei("regularGroupWithQuestionAndRegularGroupInside.xml",
                    "-1, ", "0, 0, ", "0, 1, 0, ", "0, 1, 1, ", "-1, ", "0, 1, 1, "),
            ei("regularGroupWithQuestionsAndRegularGroupInside.xml",
                    "-1, ", "0, 0, ", "0, 1, 0, ", "0, 2, ", "-1, ", "0, 2, "),
            ei("fieldListWithQuestionAndRegularGroupInside.xml",
                    "-1, ", "0, ", "-1, ", "0, "),
            ei("fieldListWithQuestionsAndRegularGroupsInside.xml",
                    "-1, ", "0, ", "-1, ", "0, "),
            ei("threeNestedFieldListGroups.xml",
                    "-1, ", "0, ", "-1, ", "0, "));
}

/** Expected indices for each form */
private static Object[] ei(String formName, String... expectedIndices) {
    return new Object[] {formName, expectedIndices};
}

```

dcbriccetti commented:
Thanks, @grzesiek2010. I made some more suggestions, none of which would change the logic if implemented, so I’ll now mark as needs testing.

mmarciniak90 commented:
**Tested with success!**

Verified on Android: 4.1, 4.2, 4.4, 5.1, 6.0, 7.0, 8.0
Verified cases:
* new forms
* saved forms
* jumping to question / go to end / go to start

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"

grzesiek2010 committed:
Improvements

lognaturel commented:
Thank for your patience through all the iterating here, @grzesiek2010!
